Unit tests #80

pedrobalage opened the pull request:
@ramon-astudillo  and @kepler,

I completed all unit tests for the school days. I started from the current develop branch and I've already merged Mariana's pull request.

I configured the Travis script. One good thing is to enable travis to check for each pull request if all the tests are passing.

All the tests are OK, but some problems were found in the toolkit:

- Python 3.5 and python 3.6 are given different results for syntax and parsing tests. I am trying to address this problem which I believe is due to a different dict implementation for python 3.6 (https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-compactdict).
- loglinear and backpropagation in numpy and pytorch also have a similar issue (different results from different versions of python).

I believe you can merge the unit tests as they will not change. I can open these new issues in github.

ramon-astudillo commented:
@pedrobalage @kepler I am having some trouble running this locally. I am using osx 10.12.1

- `tox` is not in the `requests.txt`
-  Running `tox` yields following errors both in python 2 and 3

```
tox
Successfully installed tox-2.9.1 virtualenv-15.1.0
(venv) TIEInterceptor:lxmls-toolkit ramon$ tox
py27 create: /Users/ramon/LxMLS/2018/lxmls-toolkit/.tox/py27
py27 installdeps: six, pytest, pytest-xdist, numpy, scipy, nltk, matplotlib, http://download.pytorch.org/whl/cpu/torch-0.3.1-cp27-cp27mu-linux_x86_64.whl
ERROR: invocation failed (exit code 1), logfile: /Users/ramon/LxMLS/2018/lxmls-toolkit/.tox/py27/log/py27-1.log
ERROR: actionid: py27
msg: getenv
cmdargs: ['/Users/ramon/LxMLS/2018/lxmls-toolkit/.tox/py27/bin/pip', 'install', 'six', 'pytest', 'pytest-xdist', 'numpy', 'scipy', 'nltk', 'matplotlib', 'http://download.pytorch.org/whl/cpu/torch-0.3.1-cp27-cp27mu-linux_x86_64.whl']

torch-0.3.1-cp27-cp27mu-linux_x86_64.whl is not a supported wheel on this platform.

ERROR: could not install deps [six, pytest, pytest-xdist, numpy, scipy, nltk, matplotlib, http://download.pytorch.org/whl/cpu/torch-0.3.1-cp27-cp27mu-linux_x86_64.whl]; v = InvocationError('/Users/ramon/LxMLS/2018/lxmls-toolkit/.tox/py27/bin/pip install six pytest pytest-xdist numpy scipy nltk matplotlib http://download.pytorch.org/whl/cpu/torch-0.3.1-cp27-cp27mu-linux_x86_64.whl (see /Users/ramon/LxMLS/2018/lxmls-toolkit/.tox/py27/log/py27-1.log)', 1)
py35 create: /Users/ramon/LxMLS/2018/lxmls-toolkit/.tox/py35
ERROR: InterpreterNotFound: python3.5
py36 create: /Users/ramon/LxMLS/2018/lxmls-toolkit/.tox/py36
py36 installdeps: six, pytest, pytest-xdist, numpy, scipy, nltk, matplotlib, http://download.pytorch.org/whl/cpu/torch-0.3.1-cp36-cp36m-linux_x86_64.whl
ERROR: invocation failed (exit code 1), logfile: /Users/ramon/LxMLS/2018/lxmls-toolkit/.tox/py36/log/py36-1.log
ERROR: actionid: py36
msg: getenv
cmdargs: ['/Users/ramon/LxMLS/2018/lxmls-toolkit/.tox/py36/bin/pip', 'install', 'six', 'pytest', 'pytest-xdist', 'numpy', 'scipy', 'nltk', 'matplotlib', 'http://download.pytorch.org/whl/cpu/torch-0.3.1-cp36-cp36m-linux_x86_64.whl']

torch-0.3.1-cp36-cp36m-linux_x86_64.whl is not a supported wheel on this platform.

ERROR: could not install deps [six, pytest, pytest-xdist, numpy, scipy, nltk, matplotlib, http://download.pytorch.org/whl/cpu/torch-0.3.1-cp36-cp36m-linux_x86_64.whl]; v = InvocationError('/Users/ramon/LxMLS/2018/lxmls-toolkit/.tox/py36/bin/pip install six pytest pytest-xdist numpy scipy nltk matplotlib http://download.pytorch.org/whl/cpu/torch-0.3.1-cp36-cp36m-linux_x86_64.whl (see /Users/ramon/LxMLS/2018/lxmls-toolkit/.tox/py36/log/py36-1.log)', 1)
```

ramon-astudillo commented on .travis.yml:
```diff
@@ -1,44 +1,19 @@
-    on_success: never
-    on_failure: always
+  - pip install -r requirements.txt
+  - if [[ $TRAVIS_PYTHON_VERSION == 2.7 ]]; then pip install http://download.pytorch.org/whl/cpu/torch-0.3.1-cp27-cp27mu-linux_x86_64.whl; fi
```
This line and subsequent assume `Linux`

pedrobalage replied:
The Travis is only testing now on Linux.

+os:
+  - linux

ramon-astudillo commented:
The problem is in `travis.yml`. It assumes  `Linux`.

pedrobalage commented:
Thanks Ramon. I am going to look this issue of tox on osx. 

I am afraid we could not run tests on windows too, because pytorch don't support officially windows OS. Is this bad for the school? I remembered many students with windows PCs.

Travis is independent of tox and it allow us to run the tests in travis servers. It is just a matter of activate the repository (https://travis-ci.org/LxMLS/lxmls-toolkit).

Pedro

pedrobalage commented:
@ramon-astudillo  and @kepler 

* I added new configurations in tox for testing under osx environment. 
* I also modified the tolerance parameter to allow some margin for the non-linear classifier (numpy and pytorch). 
* With this last modification, all tests are now passing on python 2.7 and 3.5.
* The tests are not passing in py 3.6 due to some strange problem with "parsing day" functions, possibly because of some buddy dict reader. I am going to raise an issue for this to be solved independently of this pull request.
* Finally, as I saw in many projects, I left the requirements.txt for using the lxmls while requirements-dev.txt for development/testing  the code.

pedrobalage commented:
After fixing #89 with the suggestion from @dcferreira, all the tests are now passing.

dcferreira commented:
@ramon-astudillo it's probably a good idea to activate Travis-CI for the repo and require that the test passes when making a pull request, since we already have a `.travis.yml`. Check https://travis-ci.org/LxMLS/lxmls-toolkit
